### PR TITLE
Implement TrySubmit in daml3-script

### LINKED ISSUE #16895: Implement TrySubmit
Part of #16862 
Should not be merged into main daml script, instead we should fork daml script (pending discussion)

Implement `trySubmit : [Party] -> [Party] -> Commands a -> Script (Either SubmitError a)`, as well as implementing `SubmitError` to cover all the cases we can achieve.

Getting errors across ledger API is quite involved, so for now, simply implement in regards to IDELedger and set up the interface in the ScriptLedgerClients.

### PR DESCRIPTION
**Pull request by samuel-williams-da**
Supersedes #16886
Resolves #16895

Adds `trySubmit`, which returns an either SubmitError or result
alongside `trySubmitMulti`

As future tickets:
- Improve `Show SubmitError` further
- Update `ContractIdInContractKey` to include `TemplateId`
- Implement lookup errors throughout, look into how canton behaves when using non existent contracts
- Implement RejectedAuthorityRequest error
- Update ContractNotFound to provide optional context, and be the catch all (#17154)

### CODE REVIEW

File: daml-script/daml3/Daml/Script/Questions/Submit/Error.daml
Comment (samuel-williams-da) [inline daml-script/daml3/Daml/Script/Questions/Submit/Error.daml:1]: From @moisesackerman-da :

I think eventually we will want an opaque representation plus extractor functions (similar to AnyTemplate + one fromAnyTemplate per type) but this is a good start and does a good job of collecting the error states

File: compiler/damlc/tests/daml-test-files/Daml3ScriptTrySubmit.daml
Comment (samuel-williams-da) [inline compiler/damlc/tests/daml-test-files/Daml3ScriptTrySubmit.daml:197]: Depending on if we want to keep Truncated error, would love ideas to resolve this

File: daml-script/daml3/Daml/Script/Questions/Submit/Error.daml
Comment (samuel-williams-da) [inline daml-script/daml3/Daml/Script/Questions/Submit/Error.daml:96]: This error type exists for when we get errors with missing metadata resources, and can't construct the full error above.
I tried making a contract key with 100,000 of the same party and I didn't get truncated.
I'm not sure how big the error even needs to get to trigger this.
Given its such a pain to even get - should we expose this? Or should we runtime error in the event of truncated errors from the ledger.

File: daml-script/daml3/Daml/Script/Questions/Submit/Error.daml
Comment (samuel-williams-da) [inline daml-script/daml3/Daml/Script/Questions/Submit/Error.daml:79]: This will be another ticket

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (samuel-williams-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:56]: Unsure how often this can happen. I can update the error in daml to take a list of contract ids, but that'll be strange for the user. I could also add a multi variant of it, but then users will likely have to handle both.
I dont know what the user should expect here, nor do I know when we'd ever see this error with multiple ids (the engine doesn't throw it)

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/IdeLedgerClient.scala
Comment (samuel-williams-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/IdeLedgerClient.scala:364]: Another ticket, following the package vetting logic, we should have a proper error for non-existent template (or any lookup failure). I'm not sure what the ledger-api does currently if you try to use a template that doesn't exist

File: daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/Daml2ScriptTestRunner.scala
Comment (samuel-williams-da) [inline daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/Daml2ScriptTestRunner.scala:1]: The DAmlScriptTestRunner was split into a trait and 2 versions, one for daml2 and one for daml3, which does all that daml2 does + any new daml3 features we feel like adding (and need testing against a real ledger).
Githubs diffs here aren't great - everything in `DamlScriptTestRunner` is old code

**Review by samuel-williams-da — COMMENTED** [review]
After this task - I'd like to combine all the submits together into a `trySubmitTree`, and handle everything on the daml side.
This will include dropping the json api.
However, the IDELedger makes this difficult, as it serialises "steps" it takes, including calls to `submitMustFail`, which won't be detectable after this refactor

File: compiler/damlc/tests/daml-test-files/Daml3ScriptTrySubmit.daml
Comment (samuel-williams-da) [reply compiler/damlc/tests/daml-test-files/Daml3ScriptTrySubmit.daml:197]: I've been a fool - we're not truncating on the IDELedger, this could only fail in canton tests.

File: daml-script/daml3/Daml/Script/Questions/Submit/Error.daml
Comment (dylant-da) [reply daml-script/daml3/Daml/Script/Questions/Submit/Error.daml:97]: For anyone skimming: https://github.com/digital-asset/daml/pull/17043/files#r1262245329

File: compiler/damlc/tests/daml-test-files/Daml3ScriptTrySubmit.daml
Comment (samuel-williams-da) [reply compiler/damlc/tests/daml-test-files/Daml3ScriptTrySubmit.daml:219]: TODO: Update this so it throws a specific error on truncation, and assert that error in the script runner test

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (dylant-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:63]: I see it is produced in `ledger/ledger-api-errors/src/main/scala/com/daml/error/definitions/groups/ConsistencyErrors.scala`, under `MultipleContractsNotFound`, but not much beyond that.

Can we default to some NonEmpty list in the `ContractNotFound` error on the daml side, rather than adding a new variant?

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/IdeLedgerClient.scala
Comment (dylant-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/IdeLedgerClient.scala:295]: Might want to leave comments drawing attention to this, so that the next person making changes knows not to leak info.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/ScriptLedgerClient.scala
Comment (dylant-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/ScriptLedgerClient.scala:220]: ~~Hello not ready, I'm dad~~

Even considering this is transient, maybe "`trySubmit` is not yet implemented in the `ScriptLedgerClient`" would be more helpful.

File: arbitrary_canton_sha
Comment (dylant-da) [inline arbitrary_canton_sha:1]: I assume this shouldn't be committed?

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/SubmitError.scala
Comment (dylant-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/SubmitError.scala:295]: Phew that's a big plate of boiler, thanks for getting this one through.

File: daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/Daml3ScriptTestRunner.scala
Comment (dylant-da) [inline daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/Daml3ScriptTestRunner.scala:43]: Do we want to change XXXXXXXX back to some correct hash?

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (remyhaemmerle-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:63]: > Unsure how often this can happen. 
please ask on #team-canton slack or during the AMA

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcLedgerClient.scala
Comment (remyhaemmerle-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcLedgerClient.scala:299]:
```suggestion
    transactionTreeF.map{ transactionTree => 
      val events = transactionTree.getTransaction.rootEventIds
        .map(evId => transactionTree.getTransaction.eventsById(evId))
        .toList
      events.traverse(fromTreeEvent(_)) match {
        case Left(err) => throw new ConverterException(err)
        case Right(results) => results
      }
    }
```

**Review by remyhaemmerle-da — APPROVED** [review]
Nice!

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:88]: In this case, the behaviour is intentional - if we cannot decode the value, we assume its because it was truncated non-gracefully such that its not parsible, so the match doesn't hit and we give back the Truncated error

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:24]: My bad - I'll fix this.

File: arbitrary_canton_sha
Comment (samuel-williams-da) [reply arbitrary_canton_sha:1]: Right, it was needed while I was waiting for a second PR to canton to merge, but that merged this morning.
Sadly I cannot drop this however until canton pushes artefacts

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/SubmitError.scala
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/SubmitError.scala:295]: Yeah - i wish it could be refactored to less, but I'm not convinced it can

File: daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/Daml3ScriptTestRunner.scala
Comment (samuel-williams-da) [reply daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/Daml3ScriptTestRunner.scala:43]: Ah, I can return them, though I'm not sure if we should be relying on them being static, doesn't seem like important information to this test

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcLedgerClient.scala
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcLedgerClient.scala:299]: The curly brackets always get me

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (remyhaemmerle-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:24]: I mistaken. Bytes.fromString use `base16`. 

It is not a problem to import guava, as protobuf uses it anyway.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (remyhaemmerle-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:88]: fair enough.

File: daml-script/daml3/Daml/Script/Questions/Submit/Error.daml
Comment (samuel-williams-da) [reply daml-script/daml3/Daml/Script/Questions/Submit/Error.daml]: Improved most of them, left a todo for things like contract keys

File: daml-script/daml3/Daml/Script/Questions/Submit/Error.daml
Comment (samuel-williams-da) [reply daml-script/daml3/Daml/Script/Questions/Submit/Error.daml:97]: Yeah I agree

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:63]: I've reimplemented this as a NonEmpty for now, multiple contract errors are thrown only by canton sandbox, but its unclear
a) what that even is
b) how often it happens
We can reduce this back if it turns out that its never used, but for now we can be sure we're not throwing away information

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala
Comment (samuel-williams-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala:25]: Right okay - I'll make the change to not swallow the error, but continue using guava

File: daml-script/test/BUILD.bazel
Comment (remyhaemmerle-da) [inline daml-script/test/BUILD.bazel:144]:
```suggestion
    tags = [
        "cpu:4",
        "exclusive",
    ],
```
Why do we need the exclusive here ?

Comment (samuel-williams-da) [reply daml-script/test/BUILD.bazel:144]: I was just trying things to see if I can make the canton setup stop timing out - it didn't work though :(
Do you have any idea why this might be happening?

Comment (remyhaemmerle-da) [reply daml-script/test/BUILD.bazel:144]: The issue is that you use `override lazy val devMode = true` but your test is not marked  with `dev-canton-test` bazel flag.

File: daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/Daml3ScriptTestRunner.scala
Comment (remyhaemmerle-da) [inline daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/Daml3ScriptTestRunner.scala:12]: Why are you running the test in  `devMode` ?

File: daml-script/test/BUILD.bazel
Comment (samuel-williams-da) [reply daml-script/test/BUILD.bazel:144]: OH I had no idea!
It was in dev mode so I could test some things and I completely forgot - you're a hero Remy

Comment (samuel-williams-da) [reply daml-script/test/BUILD.bazel]: Never mind, I forgot i need dev mode because daml3 script runs only in LF 1.dev.
If adding the dev-canton-test will fix the problem, does that mean that dev canton is currently broken?

Comment (remyhaemmerle-da) [reply daml-script/test/BUILD.bazel]: What do you need for 1.dev ?

Comment (samuel-williams-da) [reply daml-script/test/BUILD.bazel]: Nothing specific, more than script is a dev feature, so I just put it on the dev LF version - it could probably run on LF1.15, I just didn't go through the trouble of generating daml3-script for a bunch of LF versions, given I don't even know if LF1.15 will be most recent by the time daml3-script is released

Comment (remyhaemmerle-da) [reply daml-script/test/BUILD.bazel]: Fair enough.

Comment (remyhaemmerle-da) [reply daml-script/test/BUILD.bazel]: once one #17157 is merged `canton-dev-test` should work again. 
Please revert the last 2 changes and mark your test with `dev-canton-test`. 
Just keep in mind that your tests will not be running when `SKIP_DEV_CANTON_TESTS=true`

Comment (samuel-williams-da) [reply daml-script/test/BUILD.bazel]: Right okay thanks, I'll do that and wait until 17157 is merged so I can be sure the tests pass

Comment (samuel-williams-da) [reply daml-script/test/BUILD.bazel]: Oh I see, my bad - i thought the squash and merge would hide that

**Comment by samuel-williams-da** [comment]
Test fails due to wallclock mode - was this needed?

**Pull request merged**
